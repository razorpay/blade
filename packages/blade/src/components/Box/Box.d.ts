import type { CSSObject } from 'styled-components';
import type { DotNotationSpacingStringToken } from '~src/_helpers/types';

export type BoxProps = {
  display?: CSSObject['display'];
  flex?: CSSObject['flex'];
  flexWrap?: CSSObject['flexWrap'];
  flexDirection?: CSSObject['flexDirection'];
  flexGrow?: CSSObject['flexGrow'];
  alignItems?: CSSObject['alignItems'];
  alignContent?: CSSObject['alignContent'];
  justifyContent?: CSSObject['justifyContent'];
  alignSelf?: CSSObject['alignSelf'];
  overflow?: CSSObject['overflow'];
  position?: CSSObject['position'];
  top?: CSSObject['top'];
  right?: CSSObject['right'];
  bottom?: CSSObject['bottom'];
  left?: CSSObject['left'];
  paddingTop?: DotNotationSpacingStringToken | 'auto';
  paddingBottom?: DotNotationSpacingStringToken | 'auto';
  paddingLeft?: DotNotationSpacingStringToken | 'auto';
  paddingRight?: DotNotationSpacingStringToken | 'auto';
  marginTop?: DotNotationSpacingStringToken | 'auto' | number;
  marginBottom?: DotNotationSpacingStringToken | 'auto' | number;
  marginLeft?: DotNotationSpacingStringToken | 'auto' | number;
  marginRight?: DotNotationSpacingStringToken | 'auto' | number;
  width?: CSSObject['width'];
  height?: CSSObject['height'];
  gap?: DotNotationSpacingStringToken | 'auto' | number;
  minHeight?: number;
  minWidth?: number;
  maxHeight?: number;
  maxWidth?: number;
  position?: CSSObject['position'];
  transform?: CSSObject['transform'];
  top?: CSSObject['top'];
  left?: CSSObject['left'];
  right?: CSSObject['right'];
  bottom?: CSSObject['bottom'];
  background?: CSSObject['background'];
};

export { default } from './Box.web';
