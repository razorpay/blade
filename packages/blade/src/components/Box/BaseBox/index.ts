export * from './BaseBox';
export { BaseBox as default } from './BaseBox';
export { MakeValueResponsive, BaseBoxProps, SpacingValueType } from './types';
