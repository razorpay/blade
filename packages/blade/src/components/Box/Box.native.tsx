import styled from 'styled-components/native';
import type { BoxProps } from './Box.d';
import getBoxStyles from './getBoxStyles';

const Box = styled.View<BoxProps>(getBoxStyles);

export default Box;
