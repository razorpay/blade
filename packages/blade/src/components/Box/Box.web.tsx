import styled from 'styled-components';
import type { BoxProps } from './Box.d';
import getBoxStyles from './getBoxStyles';

const Box = styled.div<BoxProps>(getBoxStyles);

export default Box;
