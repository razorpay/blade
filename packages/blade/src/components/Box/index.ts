export { Box } from './Box';
export type { BoxProps } from './BaseBox/types';
