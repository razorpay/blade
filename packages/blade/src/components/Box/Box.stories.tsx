import type { Meta } from '@storybook/react';
import React from 'react';
import { getBoxArgTypes } from './BaseBox/storybookArgTypes';
import type { BoxRefType } from './BaseBox/types';
import type { BoxProps } from '.';
import { Box } from '.';
import { Text, Title } from '~components/Typography';
import StoryPageWrapper from '~utils/storybook/StoryPageWrapper';
import { LinkToStorybook } from '~utils/storybook/LinkToStorybook';
import { castWebType, isReactNative } from '~utils';
import { Button } from '~components/Button';

// Storybook renders inside iframe so by default it doesn't support scrolling to the sections.
// So we manually read location.hash of parent window and scroll to that section on load
if (window.top) {
  document.getElementById(window.top.location.hash)?.scrollIntoView();
}

const BoxStoryMeta = {
  title: 'Components/Layout Primitives (Box)/Box',
  component: Box,
  argTypes: getBoxArgTypes(),
  parameters: {
    docs: {
      page: () => (
        <StoryPageWrapper
          componentName="Box"
          componentDescription="Box Component from Layout Primitives of Blade."
          propsDescription="All Box props support responsive objects. Props marked with 💅🏼 next to their names are the props that can also be used as styled-props on other blade components. Check out styled-props documentation for more details."
        >
          <Box paddingY="spacing.5" paddingBottom="spacing.8">
            <Title size="medium">Layout Primitives Documentation</Title>
            <Text marginTop="spacing.3">
              Check Out{' '}
              <LinkToStorybook url="Components/Layout Primitives (Box)/How to Create Layouts?">
                &quot;How to Create Layouts?&quot; Docs
              </LinkToStorybook>{' '}
              for more detailed documentation of Box
            </Text>
          </Box>
        </StoryPageWrapper>
      ),
    },
  },
} as Meta<BoxProps>;

export const Default = (args: BoxProps): React.ReactElement => {
  return (
    <Box {...args}>
      <Text>Change controls to see the parameters change for the container</Text>
    </Box>
  );
};

Default.args = {
  padding: { base: 'spacing.2', m: 'spacing.10' },
  backgroundColor: 'surface.background.level1.lowContrast',
} as BoxProps;

export const Responsive = (args: BoxProps): React.ReactElement => {
  return (
    <>
      <Text>Change screen size to see flexDirection switch between row and column</Text>
      <Box {...args}>
        <Box flex="1" backgroundColor="brand.primary.500" padding="spacing.5">
          <Text contrast="high">Box1</Text>
        </Box>
        <Box flex="1" backgroundColor="surface.background.level2.highContrast" padding="spacing.5">
          <Text contrast="high">Box2</Text>
        </Box>
      </Box>
    </>
  );
};

Responsive.args = {
  display: 'flex',
  paddingY: 'spacing.6',
  flexDirection: { base: 'column', m: 'row' },
} as BoxProps;

export const AsSection = (args: BoxProps): React.ReactElement => {
  if (isReactNative()) {
    return (
      <Box>
        <Text>as prop is not supported on React Native. Check the same story on web</Text>
      </Box>
    );
  }

  return (
    <Box {...args}>
      <Text>This box is rendered as {args.as} HTML tag</Text>
    </Box>
  );
};

AsSection.args = {
  as: 'section',
} as BoxProps;

export const WithRef = (args: BoxProps): React.ReactElement => {
  const ref = React.useRef<BoxRefType>(null);

  return (
    <Box height="300px" overflow="auto" backgroundColor="surface.background.level2.lowContrast">
      <Button
        onClick={() => {
          if (!isReactNative()) {
            castWebType(ref.current).scrollIntoView();
          }
        }}
      >
        Click to Scroll
      </Button>
      <Box ref={ref} {...args}>
        <Text>Hi from Box with ref</Text>
      </Box>
    </Box>
  );
};

WithRef.args = {
  marginTop: '800px',
} as BoxProps;

export const WithMouseEvents = (args: BoxProps): React.ReactElement => {
  return (
    <Box
      {...args}
      onMouseOver={(e) => console.log('onMouseOver', e)}
      onMouseEnter={(e) => console.log('onMouseEnter', e)}
      onMouseLeave={(e) => console.log('onMouseLeave', e)}
      onScroll={(e) => console.log('onScroll', e)}
    >
      <Text marginY="300px">Move mouse over this text and check console</Text>
    </Box>
  );
};

WithMouseEvents.args = {
  overflowY: 'auto',
  height: '300px',
} as BoxProps;

export default BoxStoryMeta;
