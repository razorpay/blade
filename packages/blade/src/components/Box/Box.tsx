import React from 'react';
import BaseBox from './BaseBox';
import type { BoxProps, BoxRefType, MakeValueResponsive } from './BaseBox/types';
import { validBoxAsValues } from './BaseBox/types/propsTypes';
import type { KeysRequired } from '~utils/types';
import { isReactNative } from '~utils';
import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
import { throwBladeError } from '~utils/logger';

const validateBackgroundString = (stringBackgroundColorValue: string): void => {
  if (__DEV__) {
    if (
      !stringBackgroundColorValue.startsWith('surface.background') &&
      !stringBackgroundColorValue.startsWith('brand.') &&
      stringBackgroundColorValue !== 'transparent'
    ) {
      throwBladeError({
        message: `Oops! Currently you can only use \`transparent\`, \`surface.background.*\`, and \`brand.*\` tokens with backgroundColor property but we received \`${stringBackgroundColorValue}\` instead.\n\n Do you have a usecase of using other values? Create an issue on https://github.com/razorpay/blade repo to let us know and we can discuss âœ¨`,
        moduleName: 'Box',
      });
    }
  }
};

const validateBackgroundProp = (
  responsiveBackgroundColor: MakeValueResponsive<string | undefined>,
): void => {
  if (__DEV__) {
    if (responsiveBackgroundColor) {
      if (typeof responsiveBackgroundColor === 'string') {
        validateBackgroundString(responsiveBackgroundColor);
        return;
      }

      Object.values(responsiveBackgroundColor).forEach((backgroundColor) => {
        if (typeof backgroundColor === 'string') {
          validateBackgroundString(backgroundColor);
        }
      });
    }
  }
};

/**
 * This function is to filter out any unexpected props passed by the user
 */
const makeBoxProps = (
  props: BoxProps,
): KeysRequired<Omit<BoxProps, 'testID' | 'id' | '__brand__'>> => {
  return {
    // Layout
    display: props.display,
    overflow: props.overflow,
    overflowX: props.overflowX,
    overflowY: props.overflowY,
    whiteSpace: props.whiteSpace,
    height: props.height,
    minHeight: props.minHeight,
    maxHeight: props.maxHeight,
    width: props.width,
    minWidth: props.minWidth,
    maxWidth: props.maxWidth,
    textAlign: props.textAlign,

    // Flex
    flex: props.flex,
    flexWrap: props.flexWrap,
    flexDirection: props.flexDirection,
    flexGrow: props.flexGrow,
    flexShrink: props.flexShrink,
    flexBasis: props.flexBasis,
    alignItems: props.alignItems,
    alignContent: props.alignContent,
    alignSelf: props.alignSelf,
    justifyItems: props.justifyItems,
    justifyContent: props.justifyContent,
    justifySelf: props.justifySelf,
    placeSelf: props.placeSelf,
    placeItems: props.placeItems,
    order: props.order,

    // Grid
    grid: props.grid,
    gridColumn: props.gridColumn,
    gridRow: props.gridRow,
    gridRowStart: props.gridRowStart,
    gridRowEnd: props.gridRowEnd,
    gridColumnStart: props.gridColumnStart,
    gridColumnEnd: props.gridColumnEnd,
    gridArea: props.gridArea,
    gridAutoFlow: props.gridAutoFlow,
    gridAutoRows: props.gridAutoRows,
    gridAutoColumns: props.gridAutoColumns,
    gridTemplate: props.gridTemplate,
    gridTemplateAreas: props.gridTemplateAreas,
    gridTemplateColumns: props.gridTemplateColumns,
    gridTemplateRows: props.gridTemplateRows,

    // Spacing
    padding: props.padding,
    paddingTop: props.paddingTop,
    paddingBottom: props.paddingBottom,
    paddingRight: props.paddingRight,
    paddingLeft: props.paddingLeft,
    paddingX: props.paddingX,
    paddingY: props.paddingY,
    margin: props.margin,
    marginBottom: props.marginBottom,
    marginTop: props.marginTop,
    marginRight: props.marginRight,
    marginLeft: props.marginLeft,
    marginX: props.marginX,
    marginY: props.marginY,
    gap: props.gap,
    rowGap: props.rowGap,
    columnGap: props.columnGap,

    // Position
    position: props.position,
    zIndex: props.zIndex,
    top: props.top,
    right: props.right,
    bottom: props.bottom,
    left: props.left,

    // Visual
    backgroundColor: props.backgroundColor,
    backgroundImage: props.backgroundImage,
    backgroundSize: props.backgroundSize,
    backgroundPosition: props.backgroundPosition,
    backgroundOrigin: props.backgroundOrigin,
    backgroundRepeat: props.backgroundRepeat,
    elevation: props.elevation,
    opacity: props.opacity,
    visibility: props.visibility,

    // Border
    borderWidth: props.borderWidth,
    borderColor: props.borderColor,
    borderTopWidth: props.borderTopWidth,
    borderTopColor: props.borderTopColor,
    borderRightWidth: props.borderRightWidth,
    borderRightColor: props.borderRightColor,
    borderBottomWidth: props.borderBottomWidth,
    borderBottomColor: props.borderBottomColor,
    borderLeftWidth: props.borderLeftWidth,
    borderLeftColor: props.borderLeftColor,
    borderRadius: props.borderRadius,
    borderTopLeftRadius: props.borderTopLeftRadius,
    borderTopRightRadius: props.borderTopRightRadius,
    borderBottomRightRadius: props.borderBottomRightRadius,
    borderBottomLeftRadius: props.borderBottomLeftRadius,

    // Polygon Support
    transform: props.transform,
    transformOrigin: props.transformOrigin,
    clipPath: props.clipPath,

    // callbacks
    onMouseEnter: props.onMouseEnter,
    onMouseLeave: props.onMouseLeave,
    onMouseOver: props.onMouseOver,
    onScroll: props.onScroll,

    // Drag and Drop
    draggable: props.draggable,
    onDragStart: props.onDragStart,
    onDragEnd: props.onDragEnd,
    onDragEnter: props.onDragEnter,
    onDragLeave: props.onDragLeave,
    onDragOver: props.onDragOver,
    onDrop: props.onDrop,

    pointerEvents: props.pointerEvents,
    children: props.children,
    tabIndex: props.tabIndex,
    as: isReactNative() ? undefined : props.as, // as is not supported on react-native
  };
};

/**
 * ## Box
 * 
 * Box is the basic Layout component.
 *
 *
 * Box components supports most spacing CSS properties like `display`, `padding*`, `flex*`, `height`, `width`, etc.
 *
 *  Check out {@linkcode BoxProps BoxPropsType} for complete list of props and [Layout RFC](https://github.com/razorpay/blade/blob/master/rfcs/2023-01-06-layout.md) for more details on API decision.
 * 
 * ----
 * 
 * ### Usage
 * 
 * ```jsx
 * <Box display="flex">
 * ```

 * #### Responsive Props
 *
 * ```jsx
 * <Box padding={{ base: 'spacing.3', m: 'spacing.10' }} />
 * ```
 * 
 * #### Margin and Padding Shorthands
 * 
 * ```jsx
 * <Box padding={["spacing.3", "spacing.10"]} />
 * ```
 *
 * ---
 * 
 * Checkout {@link https://blade.razorpay.com/?path=/docs/components-box Box Documentation}
 * 
 */
const _Box: React.ForwardRefRenderFunction<BoxRefType, BoxProps> = (props, ref) => {
  React.useEffect(() => {
    if (__DEV__) {
      validateBackgroundProp(props.backgroundColor);
    }
  }, [props.backgroundColor]);

  React.useEffect(() => {
    if (__DEV__) {
      if (props.as) {
        if (isReactNative()) {
          throwBladeError({
            message: '`as` prop is not supported on React Native',
            moduleName: 'Box',
          });
        }

        if (!validBoxAsValues.includes(props.as)) {
          throwBladeError({
            message: `Invalid \`as\` prop value - ${props.as}. Only ${validBoxAsValues.join(
              ', ',
            )} are valid values`,
            moduleName: 'Box',
          });
        }
      }
    }
  }, [props.as]);

  return (
    <BaseBox
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ref={ref as any}
      id={props.id}
      {...metaAttribute({ name: MetaConstants.Box, testID: props.testID })}
      {...makeBoxProps(props)}
    />
  );
};

const Box = assignWithoutSideEffects(React.forwardRef(_Box), {
  displayName: 'Box',
  componentId: 'Box',
});

export { Box, makeBoxProps };
