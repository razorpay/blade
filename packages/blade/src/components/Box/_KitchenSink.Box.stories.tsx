import { composeStories } from '@storybook/react';
import * as boxStories from './Box.stories';
import { Box as BoxComponent } from '~components/Box';
import { Heading } from '~components/Typography';

const allStories = Object.values(composeStories(boxStories));

export const Box = (): JSX.Element => {
  return (
    <BoxComponent display="flex" flexDirection="column" gap="spacing.4">
      {allStories.map((Story) => {
        return (
          <>
            <Heading>{Story.storyName}</Heading>
            <Story />
          </>
        );
      })}
    </BoxComponent>
  );
};

export default {
  title: 'Components/KitchenSink/Box',
  component: Box,
  parameters: {
    // enable Chromatic's snapshotting only for kitchensink
    chromatic: { disableSnapshot: false },
    options: { showPanel: false },
  },
};
