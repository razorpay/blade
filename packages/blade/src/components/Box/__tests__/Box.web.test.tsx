import Box from '../Box';
import renderWithTheme from '~src/_helpers/testing/renderWithTheme.web';

describe('<Box />', () => {
  it('should render Box component with the correct styles', () => {
    const { container } = renderWithTheme(
      <Box
        display="flex"
        justifyContent="center"
        alignItems="center"
        alignSelf="center"
        paddingTop="spacing.3"
        paddingBottom="spacing.4"
        paddingLeft="spacing.5"
        paddingRight="spacing.6"
        minHeight={48}
        maxHeight={56}
        minWidth={48}
        maxWidth={56}
        position="absolute"
        top={0}
        right={0}
        bottom={0}
        left={0}
        overflow="hidden"
        background="hotpink"
        height="auto"
        width="auto"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
