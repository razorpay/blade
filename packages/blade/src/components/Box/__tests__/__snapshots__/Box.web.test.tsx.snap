// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> should render Box component with the correct styles 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  overflow: hidden;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding-top: 8px;
  padding-bottom: 12px;
  padding-left: 16px;
  padding-right: 20px;
  width: auto;
  height: auto;
  min-height: 48px;
  min-width: 48px;
  max-height: 56px;
  max-width: 56px;
  background: hotpink;
}

<div>
  <div
    class="c0"
    display="flex"
    height="auto"
    overflow="hidden"
    width="auto"
  />
</div>
`;
